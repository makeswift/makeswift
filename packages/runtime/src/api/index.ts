export * from './types'
export * from './fragments'
