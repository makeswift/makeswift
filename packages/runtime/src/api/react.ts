import { type FetchableValue } from '@makeswift/controls'
import * as MakeswiftApiClient from '../state/makeswift-api-client'
import {
  APIResourceType,
  File,
  GlobalElement,
  LocalizedGlobalElement,
  Page,
  PagePathnameSlice,
  Site,
  Snippet,
  Swatch,
  Table,
  Typography,
} from './types'
import { GraphQLClient } from './graphql/client'
import { CreateTableRecordMutation } from './graphql/documents'
import {
  CreateTableRecordMutationResult,
  CreateTableRecordMutationVariables,
} from './graphql/generated/types'

export type SerializedLocalizedResourcesMap = {
  [resourceId: string]: string | null
}

type LocalizedResourcesMap = Map<string, string | null>

export type CacheData = MakeswiftApiClient.SerializedState

export type MakeswiftClientOptions = {
  uri: string
  cacheData?: CacheData
  localizedResourcesMap?: SerializedLocalizedResourcesMap
  locale?: Intl.Locale
}

/**
 * NOTE(miguel): This "client" is used to fetch Makeswift API resources needed for the host. For
 * example, swatches, files, typographies, etc. Ideally it's internal to the runtime and is only
 * used by controls to transform API references to API resources.
 *
 * Moreover, its use should be reserved for the builder only, since for live pages all Makeswift
 * API resources should be embedded in the "page snapshot". In the builder, this client serves the
 * purpose of sending requests for API resources and keeping a cache so that changes that happen in
 * the builder, like modifying a swatch, can be sent via `postMessage` to the host and the cache can
 * immediately update the value and re-render.
 *
 * Furthermore, the API resources requested shouldn't be requested directly from the Makeswift API
 * as that would require those resources to not be authenticated since the requests come from the
 * browser when running the host. Instead, the requests should go to the host directly, at the
 * Makeswift API endpoint (i.g., `/api/makeswift/[...makeswift]` dynamic route) where the host's
 * API key can be used, securely, in the server. For this reason, this client should really be a
 * client of the host's API, not Makeswift's, intended to build and continuously maintain a realtime
 * snapshot for use in the builder, not the lives pages.
 */
export class MakeswiftHostApiClient {
  graphqlClient: GraphQLClient
  makeswiftApiClient: MakeswiftApiClient.Store
  subscribe: MakeswiftApiClient.Store['subscribe']
  private localizedResourcesMap: LocalizedResourcesMap
  private locale: Intl.Locale | undefined

  constructor({ uri, cacheData, localizedResourcesMap = {}, locale }: MakeswiftClientOptions) {
    this.graphqlClient = new GraphQLClient(uri)
    this.makeswiftApiClient = MakeswiftApiClient.configureStore({ serializedState: cacheData })
    this.subscribe = this.makeswiftApiClient.subscribe
    this.localizedResourcesMap = new Map(Object.entries(localizedResourcesMap))
    this.locale = locale
  }

  readSwatch(swatchId: string): Swatch | null {
    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.Swatch,
      swatchId,
    )
  }

  async fetchSwatch(swatchId: string): Promise<Swatch | null> {
    return await this.makeswiftApiClient.dispatch(
      MakeswiftApiClient.fetchAPIResource(APIResourceType.Swatch, swatchId),
    )
  }

  resolveSwatch(swatchId: string | undefined): FetchableValue<Swatch | null> {
    return this.resolveResource(APIResourceType.Swatch, {
      id: swatchId,
      read: id => this.readSwatch(id),
      fetch: id => this.fetchSwatch(id),
    })
  }

  readFile(fileId: string): File | null {
    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.File,
      fileId,
    )
  }

  async fetchFile(fileId: string): Promise<File | null> {
    return await this.makeswiftApiClient.dispatch(
      MakeswiftApiClient.fetchAPIResource(APIResourceType.File, fileId),
    )
  }

  resolveFile(fileId: string | undefined): FetchableValue<File | null> {
    return this.resolveResource(APIResourceType.File, {
      id: fileId,
      read: id => this.readFile(id),
      fetch: id => this.fetchFile(id),
    })
  }

  readTypography(typographyId: string): Typography | null {
    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.Typography,
      typographyId,
    )
  }

  async fetchTypography(typographyId: string): Promise<Typography | null> {
    return await this.makeswiftApiClient.dispatch(
      MakeswiftApiClient.fetchAPIResource(APIResourceType.Typography, typographyId),
    )
  }

  resolveTypography(typographyId: string | undefined): FetchableValue<Typography | null> {
    return this.resolveResource(APIResourceType.Typography, {
      id: typographyId,
      read: id => this.readTypography(id),
      fetch: id => this.fetchTypography(id),
    })
  }

  readGlobalElement(globalElementId: string): GlobalElement | null {
    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.GlobalElement,
      globalElementId,
    )
  }

  async fetchGlobalElement(globalElementId: string): Promise<GlobalElement | null> {
    return await this.makeswiftApiClient.dispatch(
      MakeswiftApiClient.fetchAPIResource(APIResourceType.GlobalElement, globalElementId),
    )
  }

  readLocalizedGlobalElement(globalElementId: string): LocalizedGlobalElement | null {
    const localizedGlobalElementId = this.getLocalizedResourceId(globalElementId)

    if (localizedGlobalElementId == null) return null

    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.LocalizedGlobalElement,
      localizedGlobalElementId,
    )
  }

  async fetchLocalizedGlobalElement(
    globalElementId: string,
  ): Promise<LocalizedGlobalElement | null> {
    const locale = this.locale

    if (locale == null) return null

    // If getLocalizedResourceId returns null, it means we have tried to fetch the resource,
    // but the resource is not available. If we haven't fetched it yet, it'll return undefined.
    const noLocalizedResource = this.getLocalizedResourceId(globalElementId) === null

    if (noLocalizedResource) return null

    // TODO(fikri): We're checking the cache here because the cache hit will fail on fetchAPIResource.
    // This is because in the cache we're saving the ID of the localizedGlobalElementId,
    // but we're reading the cache using the globalElementId.
    // This is very confusing and it can lead to more bugs. We need to refactor how we handle
    // localized global element.
    const cacheResult = this.readLocalizedGlobalElement(globalElementId)

    if (cacheResult) return cacheResult

    const result = await this.makeswiftApiClient.dispatch(
      MakeswiftApiClient.fetchAPIResource(
        APIResourceType.LocalizedGlobalElement,
        globalElementId,
        locale,
      ),
    )

    this.setLocalizedResourceId({
      resourceId: globalElementId,
      localizedResourceId: result?.id ?? null,
    })

    return result
  }

  readPagePathnameSlice(pageId: string): PagePathnameSlice | null {
    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.PagePathnameSlice,
      pageId,
    )
  }

  async fetchPagePathnameSlice(pageId: string): Promise<PagePathnameSlice | null> {
    return await this.makeswiftApiClient.dispatch(
      MakeswiftApiClient.fetchAPIResource(APIResourceType.PagePathnameSlice, pageId, this.locale),
    )
  }

  resolvePagePathnameSlice(pageId: string | undefined): FetchableValue<PagePathnameSlice | null> {
    return this.resolveResource(APIResourceType.PagePathnameSlice, {
      id: pageId,
      read: id => this.readPagePathnameSlice(id),
      fetch: id => this.fetchPagePathnameSlice(id),
    })
  }

  resolveResource<R>(
    type: APIResourceType,
    {
      id,
      read,
      fetch,
    }: {
      id: string | undefined
      read: (id: string) => R | null
      fetch: (id: string) => Promise<R | null>
    },
  ): FetchableValue<R | null> {
    const _read = () => (id != null ? read(id) : null)
    let lastValue: R | null = null
    return {
      name: `${type}:${id}`,
      readStable: () => (lastValue = _read()),
      subscribe: (onUpdate: () => void) =>
        this.subscribe(() => {
          if (_read() !== lastValue) onUpdate()
        }),
      fetch: async () => (id != null ? fetch(id) : null),
    }
  }

  readTable(tableId: string): Table | null {
    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.Table,
      tableId,
    )
  }

  async fetchTable(tableId: string): Promise<Table | null> {
    return await this.makeswiftApiClient.dispatch(
      MakeswiftApiClient.fetchAPIResource(APIResourceType.Table, tableId),
    )
  }

  async createTableRecord(tableId: string, columns: any): Promise<void> {
    await this.graphqlClient.request<
      CreateTableRecordMutationResult,
      CreateTableRecordMutationVariables
    >(CreateTableRecordMutation, { input: { data: { tableId, columns } } })
  }

  readSite(siteId: string): Site | null {
    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.Site,
      siteId,
    )
  }

  readPage(pageId: string): Page | null {
    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.Page,
      pageId,
    )
  }

  readSnippet(snippetId: string): Snippet | null {
    return MakeswiftApiClient.getAPIResource(
      this.makeswiftApiClient.getState(),
      APIResourceType.Snippet,
      snippetId,
    )
  }

  private getLocalizedResourceId(resourceId: string): string | undefined | null {
    return this.localizedResourcesMap?.get(resourceId)
  }

  setLocalizedResourceId({
    resourceId,
    localizedResourceId,
  }: {
    resourceId: string
    localizedResourceId: string | null
  }): void {
    this.localizedResourcesMap.set(resourceId, localizedResourceId)
  }
}
