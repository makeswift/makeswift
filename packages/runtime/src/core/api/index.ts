/**
 * Re-export all API-related types and utilities
 */
export * from './client';
export * from './types';