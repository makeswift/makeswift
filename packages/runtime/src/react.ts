export type { Action } from './state/actions'
export { ActionTypes, changeDocument } from './state/actions'
export {
  Element,
  ReactRuntime,
  useIsInBuilder,
  DocumentRoot,
  PageProvider,
  usePageId,
} from './runtimes/react'
