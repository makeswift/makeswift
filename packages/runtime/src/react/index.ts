export { ReactRuntime } from '../runtimes/react'
