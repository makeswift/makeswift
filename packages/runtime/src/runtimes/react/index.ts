export { ReactRuntime } from './react-runtime'

export { Element } from './components/Element'
export { DocumentRoot } from './components/DocumentRoot'
export { PageProvider } from './components/PageProvider'
export { ReactRuntimeProvider } from './components/RuntimeProvider'

export { useBreakpoints } from './hooks/use-breakpoints'
export { useBuilderEditMode } from './hooks/use-builder-edit-mode'
export { useDocumentKey } from './hooks/use-document-context'
export { useElementId } from './hooks/use-element-id'
export { useIsInBuilder } from './hooks/use-is-in-builder'
export { useIsPreview } from './hooks/use-is-preview'
export { usePageId } from './hooks/use-page-id'
export { useSelector } from './hooks/use-selector'
export { StoreContext, useStore } from './hooks/use-store'
