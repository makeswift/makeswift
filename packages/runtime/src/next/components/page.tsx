'use client'

import { Suspense, memo, useMemo } from 'react'

import { type CacheData } from '../../api/react'

import { useDispatch } from '../../runtimes/react/hooks/use-dispatch'
import { useUniversalDispatch } from '../../runtimes/react/hooks/use-universal-dispatch'
import { useMakeswiftHostApiClient } from '../../runtimes/react/host-api-client'

import { Page as PageComponent } from '../../components/page'
import {
  type MakeswiftPageSnapshot,
  type MakeswiftPageDocument,
  pageToRootDocument,
} from '../client'

import * as ReactPage from '../../state/react-page'
import { registerDocumentsEffect, updateAPIClientCache } from '../../state/actions'

export type PageProps = {
  snapshot: MakeswiftPageSnapshot
}

export function useCacheData(cacheData: CacheData) {
  const { makeswiftApiClient: apiStore } = useMakeswiftHostApiClient()
  useUniversalDispatch(apiStore.dispatch, updateAPIClientCache, [cacheData])
}

function useRegisterPageDocument(pageDocument: MakeswiftPageDocument): ReactPage.Document {
  const dispatch = useDispatch()
  const rootDocuments: [ReactPage.Document] = useMemo(
    () => [pageToRootDocument(pageDocument)],
    [pageDocument],
  )

  useUniversalDispatch(dispatch, registerDocumentsEffect, [rootDocuments])

  return rootDocuments[0]
}

export const Page = memo(({ snapshot, ...props }: PageProps) => {
  if ('runtime' in props) {
    throw new Error(
      `The \`runtime\` prop is no longer supported in the \`@makeswift/runtime\` \`Page\` component as of \`0.15.0\`.
See our docs for more information on what's changed and instructions to migrate: https://docs.makeswift.com/migrations/0.15.0`,
    )
  }

  useCacheData(snapshot.cacheData)

  const rootDocument = useRegisterPageDocument(snapshot.document)

  return (
    <Suspense>
      {/* We use a key here to reset the Snippets state in the PageMeta component */}
      <PageComponent
        key={snapshot.document.data.key}
        page={snapshot.document}
        rootDocument={rootDocument}
      />
    </Suspense>
  )
})
