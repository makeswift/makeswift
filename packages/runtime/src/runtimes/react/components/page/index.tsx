import { useMemo, memo, type ComponentProps, Fragment } from 'react'

import { useCacheData } from '../../hooks/use-cache-data'

import { type MakeswiftPageSnapshot, pageToRootDocument } from '../../../../client'

import { type PageMetadataSettings } from './page-seo-settings'
import { Page as PageComponent } from './Page'
import { useRegisterDocument } from '../../hooks/use-register-document'

export { type PageMetadataSettings } from './page-seo-settings'

/**
 * @param snapshot - The snapshot of the page to render, from
 * `client.getPageSnapshot()`.
 * @param metadata - Allows control over whether to use data from Makeswift for
 * rendering metadata tags in the `<head>` of the page. Pass `true` (default if
 * not provided) to render all metadata tags, or `false` to not render any. For
 * more granular control, pass an object with boolean values for specific
 * metadata fields. Valid fields include:
 *  - `title`
 *  - `description`
 *  - `keywords`
 *  - `socialImage`
 *  - `canonicalUrl`
 *  - `indexingBlocked`
 *  - `favicon`
 *
 * If a field is not provided, it will default to `false`.
 */
export const Page = memo(
  ({
    snapshot,
    metadata = true,
  }: {
    snapshot: MakeswiftPageSnapshot
    metadata?: boolean | PageMetadataSettings
  }) => {
    useCacheData(snapshot.cacheData)

    const rootDocument = useMemo(() => pageToRootDocument(snapshot.document), [snapshot.document])
    useRegisterDocument(rootDocument)

    return (
      <Fragment>
        {/* We use a key here to reset the Snippets state in the PageMeta component */}
        <PageComponent
          key={snapshot.document.data.key}
          page={snapshot.document}
          rootDocument={rootDocument}
          metadata={metadata}
        />
      </Fragment>
    )
  },
)

export type PageProps = ComponentProps<typeof Page>
