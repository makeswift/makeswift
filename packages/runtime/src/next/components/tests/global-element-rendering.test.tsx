/** @jest-environment jsdom */

import { act } from 'react-dom/test-utils'
import { render } from '@testing-library/react'
import '@testing-library/jest-dom'

import { APIResourceType, type LocalizedGlobalElement, type GlobalElement } from '../../../api'
import { ReactRuntime } from '../../../react'
import { ReactRuntimeProvider } from '../../../runtimes/react'
import { createMakeswiftPageSnapshot } from '../../../utils/tests/element-data-test-test'

import { Page } from '../page'

const globalElementId = 'R2xvYmFsRWxlbWVudDpjMzM2MDJiZC00NmFmLTQ0ZWMtODAxZi0zODVmMTEzY2Q1NzA='
const localizedGlobalElementId =
  'R2xvYmFsRWxlbWVudDpjMzM2MDJiZC00NmFmLTQ0ZWMtODAxZi0zODVmMTEzY2Q1NzA='

const swatchId = 'U3dhdGNoOjNhN2ViYWFkLTZkMTMtNDRiZC1hZDQxLWQ1NDNlMGQwMDRlMQ=='

const pageElementData = {
  key: '31bf33d5-7cde-495e-a38e-c734ba8de75e',
  props: {
    children: {
      '@@makeswift/type': 'prop-controllers::grid::v1',
      value: {
        columns: [
          {
            deviceId: 'desktop',
            value: {
              count: 12,
              spans: [[12]],
            },
          },
        ],
        elements: [
          {
            key: '71394e9c-92cd-400f-b8ba-4d4cd12dc603',
            type: 'reference',
            value: globalElementId,
          },
        ],
      },
    },
  },
  type: './components/Root/index.js',
}

const globalElementData = {
  key: '2c3d5c2a-ea20-4bc9-af04-3ce3238b4c73',
  type: './components/Text/index.js',
  props: {
    id: { value: 'title', '@@makeswift/type': 'prop-controllers::element-id::v1' },
    text: {
      key: '79d65529-9b89-4415-ba7a-dec2972d6065',
      type: 'makeswift::controls::rich-text-v2',
      version: 2,
      descendants: [
        {
          type: 'default',
          children: [
            {
              text: 'About Us',
              typography: {
                style: [
                  { value: { fontSize: { unit: 'px', value: 16 } }, deviceId: 'mobile' },
                  {
                    value: {
                      italic: false,
                      fontSize: { unit: 'px', value: 30 },
                      fontFamily: 'Lato',
                      fontWeight: 700,
                      lineHeight: 1.5,
                    },
                    deviceId: 'desktop',
                  },
                ],
              },
            },
          ],
        },
      ],
    },
    width: [{ value: { unit: 'px', value: 700 }, deviceId: 'desktop' }],
    margin: {
      value: [
        {
          value: {
            marginTop: { unit: 'px', value: 40 },
            marginLeft: 'auto',
            marginRight: 'auto',
            marginBottom: { unit: 'px', value: 20 },
          },
          deviceId: 'desktop',
        },
      ],
      '@@makeswift/type': 'prop-controllers::margin::v1',
    },
  },
} as GlobalElement['data']

const localizedGlobalElementData = {
  key: '2c3d5c2a-ea20-4bc9-af04-3ce3238b4c73',
  type: './components/Text/index.js',
  props: {
    id: { value: 'title', '@@makeswift/type': 'prop-controllers::element-id::v1' },
    text: {
      key: '9e7388b1-b743-4c2e-97ca-2af6ddca304f',
      type: 'makeswift::controls::rich-text-v2',
      version: 2,
      descendants: [
        {
          type: 'default',
          children: [
            {
              text: 'Ã€ propos de nous',
              typography: {
                style: [
                  { value: { fontSize: { unit: 'px', value: 16 } }, deviceId: 'mobile' },
                  {
                    value: {
                      color: {
                        alpha: 1,
                        swatchId,
                      },
                      italic: false,
                      fontSize: { unit: 'px', value: 30 },
                      fontFamily: 'Lato',
                      fontWeight: 700,
                      lineHeight: 1.5,
                    },
                    deviceId: 'desktop',
                  },
                ],
              },
            },
          ],
        },
      ],
    },
    width: [{ value: { unit: 'px', value: 700 }, deviceId: 'desktop' }],
    margin: {
      value: [
        {
          value: {
            marginTop: { unit: 'px', value: 40 },
            marginLeft: 'auto',
            marginRight: 'auto',
            marginBottom: { unit: 'px', value: 20 },
          },
          deviceId: 'desktop',
        },
      ],
      '@@makeswift/type': 'prop-controllers::margin::v1',
    },
  },
} as LocalizedGlobalElement['data']

async function testGlobalElementRendering({ locale }: { locale: string | null }) {
  const runtime = new ReactRuntime()
  const snapshot = createMakeswiftPageSnapshot(pageElementData, {
    locale,
    cacheData: {
      apiResources: {
        [APIResourceType.GlobalElement]: [
          {
            id: globalElementId,
            value: {
              __typename: APIResourceType.GlobalElement,
              id: globalElementId,
              data: globalElementData,
            },
          },
        ],
        [APIResourceType.LocalizedGlobalElement]: [
          {
            id: localizedGlobalElementId,
            value: {
              __typename: APIResourceType.LocalizedGlobalElement,
              id: localizedGlobalElementId,
              data: localizedGlobalElementData,
            },
            locale: 'fr-FR',
          },
        ],
        [APIResourceType.Swatch]: [
          {
            id: swatchId,
            value: {
              __typename: APIResourceType.Swatch,
              id: swatchId,
              hue: 238,
              saturation: 87,
              lightness: 49,
            },
          },
        ],
      },
      localizedResourcesMap: {
        'fr-FR': {
          [globalElementId]: localizedGlobalElementId,
        },
      },
    },
  })

  // Assert
  await act(async () =>
    render(
      <ReactRuntimeProvider runtime={runtime}>
        <Page snapshot={snapshot} />
      </ReactRuntimeProvider>,
    ),
  )

  expect(document.querySelector('#title')).toMatchSnapshot()
}

describe('Page', () => {
  test('correctly renders a global element', async () => {
    await testGlobalElementRendering({
      locale: null,
    })
  })

  test('correctly renders a localized global element', async () => {
    await testGlobalElementRendering({
      locale: 'fr-FR',
    })
  })
})
