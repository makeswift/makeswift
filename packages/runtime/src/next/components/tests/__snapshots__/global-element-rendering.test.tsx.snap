// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page correctly renders a global element 1`] = `
.emotion-0 {
  max-width: 100%;
}

@media only screen and (min-width: 769px) {
  .emotion-0 {
    width: 700px;
  }
}

@media only screen and (min-width: 576px) and (max-width: 768px) {
  .emotion-0 {
    width: 700px;
  }
}

@media only screen and (max-width: 575px) {
  .emotion-0 {
    width: 700px;
  }
}

@media only screen and (min-width: 769px) {
  .emotion-0 {
    margin-top: 40px;
    margin-right: auto;
    margin-bottom: 20px;
    margin-left: auto;
  }
}

@media only screen and (min-width: 576px) and (max-width: 768px) {
  .emotion-0 {
    margin-top: 40px;
    margin-right: auto;
    margin-bottom: 20px;
    margin-left: auto;
  }
}

@media only screen and (max-width: 575px) {
  .emotion-0 {
    margin-top: 40px;
    margin-right: auto;
    margin-bottom: 20px;
    margin-left: auto;
  }
}

.emotion-1 {
  margin: 0;
}

@media only screen and (min-width: 769px) {
  .emotion-2 {
    font-family: Lato;
    font-size: 30px;
    font-weight: 700;
    line-height: 1.5;
    font-style: initial;
  }
}

@media only screen and (min-width: 576px) and (max-width: 768px) {
  .emotion-2 {
    font-family: Lato;
    font-size: 30px;
    font-weight: 700;
    line-height: 1.5;
    font-style: initial;
  }
}

@media only screen and (max-width: 575px) {
  .emotion-2 {
    font-family: Lato;
    font-size: 16px;
    font-weight: 700;
    line-height: 1.5;
    font-style: initial;
  }
}

<div
  class="emotion-0"
  id="title"
>
  <div
    style="position: relative; white-space: pre-wrap; word-wrap: break-word;"
  >
    <p
      class="emotion-1"
    >
      <span
        class="undefined emotion-2"
      >
        About Us
      </span>
    </p>
  </div>
</div>
`;

exports[`Page correctly renders a localized global element 1`] = `
.emotion-0 {
  max-width: 100%;
}

@media only screen and (min-width: 769px) {
  .emotion-0 {
    width: 700px;
  }
}

@media only screen and (min-width: 576px) and (max-width: 768px) {
  .emotion-0 {
    width: 700px;
  }
}

@media only screen and (max-width: 575px) {
  .emotion-0 {
    width: 700px;
  }
}

@media only screen and (min-width: 769px) {
  .emotion-0 {
    margin-top: 40px;
    margin-right: auto;
    margin-bottom: 20px;
    margin-left: auto;
  }
}

@media only screen and (min-width: 576px) and (max-width: 768px) {
  .emotion-0 {
    margin-top: 40px;
    margin-right: auto;
    margin-bottom: 20px;
    margin-left: auto;
  }
}

@media only screen and (max-width: 575px) {
  .emotion-0 {
    margin-top: 40px;
    margin-right: auto;
    margin-bottom: 20px;
    margin-left: auto;
  }
}

.emotion-1 {
  margin: 0;
}

@media only screen and (min-width: 769px) {
  .emotion-2 {
    color: hsla(238,87%,49%,1);
    font-family: Lato;
    font-size: 30px;
    font-weight: 700;
    line-height: 1.5;
    font-style: initial;
  }
}

@media only screen and (min-width: 576px) and (max-width: 768px) {
  .emotion-2 {
    color: hsla(238,87%,49%,1);
    font-family: Lato;
    font-size: 30px;
    font-weight: 700;
    line-height: 1.5;
    font-style: initial;
  }
}

@media only screen and (max-width: 575px) {
  .emotion-2 {
    color: hsla(238,87%,49%,1);
    font-family: Lato;
    font-size: 16px;
    font-weight: 700;
    line-height: 1.5;
    font-style: initial;
  }
}

<div
  class="emotion-0"
  id="title"
>
  <div
    style="position: relative; white-space: pre-wrap; word-wrap: break-word;"
  >
    <p
      class="emotion-1"
    >
      <span
        class="undefined emotion-2"
      >
        Ã€ propos de nous
      </span>
    </p>
  </div>
</div>
`;
